samcli/lib/build/workflow_config.py

```diff
@@ -2,60 +2,158 @@
+                executable_search_paths=None)
 
 RUBY_BUNDLER_CONFIG = CONFIG(
                 language="ruby",
                 dependency_manager="bundler",
                 application_framework=None,
-                manifest_name="Gemfile")
+                manifest_name="Gemfile",
+                executable_search_paths=None)
+
+JAVA_GRADLE_CONFIG = CONFIG(
+                language="java",
+                dependency_manager="gradle",
+                application_framework=None,
+                manifest_name="build.gradle",
+                executable_search_paths=None)
 
 
 class UnsupportedRuntimeException(Exception):
     pass
 
 
-def get_workflow_config(runtime):
+def get_workflow_config(runtime, code_dir, project_dir):
     """
-    Get a workflow config that corresponds to the runtime provided
+    Get a workflow config that corresponds to the runtime provided. This method examines contents of the project
+    and code directories to determine the most appropriate workflow for the given runtime. Currently the decision is
+    based on the presence of a supported manifest file. For runtimes that have more than one workflow, we choose a
+    workflow by examining ``code_dir`` followed by ``project_dir`` for presence of a supported manifest.
 
     Parameters
     ----------
     runtime str
         The runtime of the config
 
+    code_dir str
+        Directory where Lambda function code is present
+
+    project_dir str
+        Root of the Serverless application project.
+
     Returns
     -------
     namedtuple(Capability)
         namedtuple that represents the Builder Workflow Config
     """
 
-    workflow_config_by_runtime = {
-        "python2.7": PYTHON_PIP_CONFIG,
-        "python3.6": PYTHON_PIP_CONFIG,
-        "python3.7": PYTHON_PIP_CONFIG,
-        "nodejs4.3": NODEJS_NPM_CONFIG,
-        "nodejs6.10": NODEJS_NPM_CONFIG,
-        "nodejs8.10": NODEJS_NPM_CONFIG,
-        "ruby2.5": RUBY_BUNDLER_CONFIG
+    selectors_by_runtime = {
+        "python2.7": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "python3.6": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "python3.7": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "nodejs4.3": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "nodejs6.10": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "nodejs8.10": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "ruby2.5": BasicWorkflowSelector(RUBY_BUNDLER_CONFIG),
+
+        # When Maven builder exists, add to this list so we can automatically choose a builder based on the supported
+        # manifest
+        "java8": ManifestWorkflowSelector([
+            # Gradle builder needs custom executable paths to find `gradlew` binary
+            JAVA_GRADLE_CONFIG._replace(executable_search_paths=[code_dir, project_dir])
+        ]),
     }
 
-    try:
-        return workflow_config_by_runtime[runtime]
-    except KeyError:
+    if runtime not in selectors_by_runtime:
```

Comment:
It's much cleaner if you do

```
try:
  selector = selectors_by_runtime[runtime]
  config = selector.get_config(code_dir, project_dir)
except KeyError:
  #blah 
except ValueError:
  #another blah 
```

Also more pythonic :)

Reply:
Not really. try/catch blocks should be scoped very tightly to the code you expect to throw. So when you know two statemtents will throw for two different reasons, I would rather have them in separte try blocks or let them throw very specific exceptions - not KeyError & ValueError. That's why I made this way.